java/spark/src/main/java/com/logicalclocks/hsfs/spark/engine/hudi/DeltaStreamerKafkaSource.java

```diff
@@ -75,39 +85,43 @@ public DeltaStreamerKafkaSource(TypedProperties props, JavaSparkContext sparkCon
         throw new HoodieException(error, var9);
       }
     }
-
-    HoodieWriteConfig.Builder builder = HoodieWriteConfig.newBuilder();
-    this.metrics = new HoodieDeltaStreamerMetrics(builder.withProperties(props).build());
-    this.offsetGen = new KafkaOffsetGen(props);
+    offsetGen = new KafkaOffsetGen(props);
   }
-
-  protected InputBatch<JavaRDD<GenericRecord>> fetchNewData(Option<String> lastCheckpointStr, long sourceLimit) {
-    OffsetRange[] offsetRanges = this.offsetGen.getNextOffsetRanges(lastCheckpointStr, sourceLimit, this.metrics);
-    long totalNewMsgs = KafkaOffsetGen.CheckpointUtils.totalNewMessages(offsetRanges);
-    LOG.info("About to read " + totalNewMsgs + " from Kafka for topic: " + this.offsetGen.getTopicName()
-        + " from offsets: " + Arrays.stream(offsetRanges)
-            .map(r -> String.format("%s:%d", r.partition(), r.fromOffset())).collect(Collectors.joining(","))
-        + " until offsets: " + Arrays.stream(offsetRanges)
-            .map(r -> String.format("%s:%d", r.partition(), r.untilOffset())).collect(Collectors.joining(",")));
-    if (totalNewMsgs <= 0L) {
-      return new InputBatch(Option.empty(), KafkaOffsetGen.CheckpointUtils.offsetsToStr(offsetRanges));
+  
+  
+  @Override
+  protected InputBatch<JavaRDD<GenericRecord>> readFromCheckpoint(Option<Checkpoint> lastCheckpointStr,
+      long sourceLimit) {
+    return super.readFromCheckpoint(lastCheckpointStr, sourceLimit);
+  }
+  
+  protected JavaRDD<GenericRecord> maybeAppendKafkaOffsets(JavaRDD<ConsumerRecord<Object, Object>> kafkaRDd) {
+    if (shouldAddOffsets) {
+      AvroConvertor convertor = new AvroConvertor(schemaProvider.getSourceSchema());
+      return kafkaRDd.map(convertor::withKafkaFieldsAppended);
     } else {
-      JavaRDD<GenericRecord> newDataRdd = this.toRdd(offsetRanges);
-      return new InputBatch(Option.of(newDataRdd), KafkaOffsetGen.CheckpointUtils.offsetsToStr(offsetRanges));
+      return kafkaRDd.map(consumerRecord -> (GenericRecord) consumerRecord.value());
     }
   }
-
-  private JavaRDD<GenericRecord> toRdd(OffsetRange[] offsetRanges) {
-    return KafkaUtils.createRDD(this.sparkContext, this.offsetGen.getKafkaParams(), offsetRanges,
-        LocationStrategies.PreferConsistent())
-            .filter(obj -> obj.value() != null)
-            .map(obj -> (GenericRecord) obj.value());
+  
+  @Override
+  protected JavaRDD<GenericRecord> toBatch(OffsetRange[] offsetRanges) {
+    JavaRDD<ConsumerRecord<Object, Object>> kafkaRDd;
+    LOG.info("Kafka offset ranges: " + offsetRanges.length);
```

Comment:
previously we did:
```
long totalNewMsgs = KafkaOffsetGen.CheckpointUtils.totalNewMessages(offsetRanges);
    LOG.info("About to read " + totalNewMsgs + " from Kafka for topic: " + this.offsetGen.getTopicName()
        + " from offsets: " + Arrays.stream(offsetRanges)
            .map(r -> String.format("%s:%d", r.partition(), r.fromOffset())).collect(Collectors.joining(","))
        + " until offsets: " + Arrays.stream(offsetRanges)
            .map(r -> String.format("%s:%d", r.partition(), r.untilOffset())).collect(Collectors.joining(",")));
```
I dont think this implementation is as useful (it just shows `offsetRanges.length`)